File: main/n-ary-tree-postorder-traversal/n-ary-tree-postorder-traversal.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution(object):
    def postorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        ret = []
        if root is None:
            return ret

        stack = [(iter(root.children), root.val)]

        while stack:
            it, val = stack[-1]

            try:
                child = next(it)
                stack.append((iter(child.children), child.val))
            except StopIteration:
                ret.append(val)
                del stack[-1]

        return ret

File: main/n-ary-tree-postorder-traversal/n-ary-tree-postorder-traversal.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> children;

    public Node(){}
    public Node(int _val,IList<Node> _children) {
        val = _val;
        children = _children;
}
*/
public class Solution {
    public IList<int> Postorder(Node root) {
        var ret = new List<int>();
        if (root == null) return ret;

        var stack = new Stack<Frame>();
        stack.Push(new Frame(root));

        while (stack.Count != 0) {
            var frame = stack.Peek();

            if (frame.Index == frame.Parent.children.Count) {
                ret.Add(frame.Parent.val);
                stack.Pop();
            } else {
                stack.Push(new Frame(frame.Parent.children[frame.Index++]));
            }
        }

        return ret;
    }

    private sealed class Frame {
        internal Frame(Node parent) => Parent = parent;

        internal Node Parent { get; }
        internal int Index { get; set; } = 0;
    }
}
